devdoc/waiter.md

```diff
@@ -0,0 +1,231 @@
+```c
+#define WAITER_RESULT_VALUES \
+    WAITER_RESULT_SYNC, \
+    WAITER_RESULT_ASYNC, \
+    WAITER_RESULT_REFUSED, \
+    WAITER_RESULT_ERROR
+
+MU_DEFINE_ENUM(WAITER_RESULT, WAITER_RESULT_VALUES);
+
+#define WAITER_CALLBACK_RESULT_VALUES \
+    WAITER_CALLBACK_RESULT_OK, \
+    WAITER_CALLBACK_RESULT_CANCELLED, \
+    WAITER_CALLBACK_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(WAITER_CALLBACK_RESULT, WAITER_CALLBACK_RESULT_VALUES);
+
+#include "umock_c/umock_c_prod.h"
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+typedef void(*NOTIFICATION_CALLBACK)(void* context, THANDLE(RC_PTR) data, WAITER_CALLBACK_RESULT result);
+typedef void(*NOTIFY_COMPLETE_CALLBACK)(void* context, WAITER_CALLBACK_RESULT result);
+typedef struct WAITER_TAG WAITER;
+
+THANDLE_TYPE_DECLARE(WAITER);
+
+    MOCKABLE_FUNCTION(, THANDLE(WAITER), waiter_create)
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_register_notification, THANDLE(WAITER), waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_notify, THANDLE(WAITER), waiter, THANDLE(RC_PTR), data, NOTIFY_COMPLETE_CALLBACK, notify_complete_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### waiter_create
+```c
+    MOCKABLE_FUNCTION(, WAITER_HANDLE, waiter_create)
+```
+
+`waiter_create` creates the waiter and returns it.
+
+**SRS_WAITER_43_001: [** `waiter_create` shall call `THANDLE_MALLOC` with `dispose` as `waiter_dispose`. **]**
+
+**SRS_WAITER_43_002: [** If there are any failures, `waiter_create` shall fail and return `NULL`. **]**
+
+### waiter_dispose
+```c
+    void waiter_dispose(WAITER* waiter)
+```
+
+**SRS_WAITER_43_003: [** `waiter_dispose` disposes the given `waiter`. **]**
+
+**SRS_WAITER_43_004: [** `waiter_dispose` shall obtain the current state of the waiter by calling `interlocked_and`. **]**
+
+**SRS_WAITER_43_005: [** If the `REGISTERED` bit is set, `waiter_dispose` shall call the `notification_callback` given to `waiter_register_notification` with `context` as the `context` given to `waiter_register_notification`, `data` as `NULL` and `result` as `WAITER_RESULT_ABANDONED`. **]**
+
+**SRS_WAITER_43_006: [** If the `NOTIFIED` bit is set, `waiter_dispose` shall: **]**
+
+- **SRS_WAITER_43_044: [** call the `notify_complete_callback` given to `waiter_notify` with `context` as the `context` given to `waiter_notify`, and `result` as `WAITER_RESULT_ABANDONED`. **]**
+
+- **SRS_WAITER_43_045: [** release the reference to `data` stored in `waiter` by calling `THANDLE_ASSIGN`. **]**
+
+### waiter_register_notification
+```c
+    MOCKABLE_FUNCTION(, WAITER_RESULT, waiter_register_notification, THANDLE(WAITER), waiter, NOTIFICATION_CALLBACK, notification_callback, void*, context, THANDLE(ASYNC_OP)*, out_op);
+```
+
+`waiter_register_notification` registers the given `notification_callback` to be called when `waiter_notify` is called.
```

Comment:
>when `waiter_notify` is called [](http://example.com/codeflow?start=88&length=30)

hmmmm. Consider adding here some explanatory wording around "when data has been consumed/is preparing to be consumed"...